From 72dd30d92955b31d61536f432d5e98d9bbb35b40 Mon Sep 17 00:00:00 2001
From: Longwu Ou <xulongwu4@gmail.com>
Date: Sun, 3 Jan 2021 17:45:19 -0500
Subject: [PATCH] Add support for stateless configuration

---
 src/main.go | 21 +++++++++++++++++----
 1 file changed, 17 insertions(+), 4 deletions(-)

diff --git a/src/main.go b/src/main.go
index 8da5874..e8ffb3b 100644
--- a/src/main.go
+++ b/src/main.go
@@ -20,10 +20,9 @@ const (
 	configFilePath      = "nvidia-container-runtime/config.toml"
 
 	hookDefaultFilePath = "/usr/bin/nvidia-container-runtime-hook"
-)
 
-var (
-	configDir           = "/etc/"
+	sysConfigDir        = "/etc/"
+	vendorConfigDir     = "/usr/share/"
 )
 
 var fileLogger *log.Logger = nil
@@ -39,11 +38,25 @@ type config struct {
 
 func getConfig() (*config, error) {
 	cfg := &config{}
+    var configDir string
 
+    // First look for config file in XDG_CONFIG_HOME
 	if XDGConfigDir := os.Getenv(configOverride); len(XDGConfigDir) != 0 {
-		configDir = XDGConfigDir
+        if _, err := os.Stat(path.Join(XDGConfigDir, configFilePath)); err == nil {
+		    configDir = XDGConfigDir
+        }
 	}
 
+    // If no config file in XDG_CONFIG_HOME, load /etc/ first, otherwise fallback to
+    // /usr/share/ for stateless operation
+    if len(configDir) == 0 {
+        if _, err := os.Stat(path.Join(sysConfigDir, configFilePath)); err == nil {
+		    configDir = sysConfigDir
+        } else {
+            configDir = vendorConfigDir
+        }
+    }
+
 	configFilePath := path.Join(configDir, configFilePath)
 
 	tomlContent, err := ioutil.ReadFile(configFilePath)
-- 
2.29.2

